tests/migrations/test_loader.py

```diff
@@ -81,6 +81,24 @@ def test_load(self):
         # Ensure we've included unmigrated apps in there too
         self.assertIn("basic", project_state.real_apps)
 
+    @override_settings(MIGRATION_MODULES={
+        "migrations": "migrations.test_migrations",
+        "migrations2": "migrations2.test_migrations_2",
+    })
+    @modify_settings(INSTALLED_APPS={'append': 'migrations2'})
+    def test_plan_handles_repeated_migrations(self):
+        """
+        _generate_plan does not re-add migrations already in plan.
+        Regression test for #29180.
+        """
+        migration_loader = MigrationLoader(connection)
+        graph = migration_loader.graph
+        nodes = [("migrations", "0002_second"), ("migrations2", "0001_initial")]
+        self.assertEqual(
+            graph._generate_plan(nodes, True),
```

Comment:
`migration_loader.graph._generate_plan(nodes, end_at=True),`
(omit intermediate `graph` var and pass True as kwarg for readability.